### PR TITLE
Update haproxy and add checks for services

### PR DESCRIPTION
**Pull request by aledbf**

### CODE REVIEW

File: service-loadbalancer/service_loadbalancer.go [hunk 1]
```diff
@@ -140,6 +140,8 @@ type service struct {
 	Name string
 	Ep   []string
 
+	BackendPort int
```
Comment (bprashanth) [inline service-loadbalancer/service_loadbalancer.go]: please add a comment, something like: 
```
Kubernetes endpoint port. The application must serve a 200 page on this port.
```

File: service-loadbalancer/service_loadbalancer.go [hunk 2]
```diff
@@ -388,8 +390,9 @@ func (lbc *loadBalancerController) getServices() (httpSvc []service, tcpSvc []se
 				continue
 			}
 			newSvc := service{
-				Name: getServiceNameForLBRule(&s, servicePort.Port),
-				Ep:   ep,
+				Name:        getServiceNameForLBRule(&s, servicePort.Port),
+				Ep:          ep,
+				BackendPort: servicePort.Port,
```
Comment (bprashanth) [inline service-loadbalancer/service_loadbalancer.go]: You're going to use this for health checks on the endpoint right? don't you want the endpoint port? I mean, you can have an nginx service with port 80, and targetport 8080 going to an endpoint running on port 8080.

File: service-loadbalancer/haproxy_reload [hunk 3]
```diff
@@ -23,6 +23,4 @@
 # -s soft reload, wait for pids to finish handling requests
 # -f send pids a resume signal if reload of new config fails
 
-socat /tmp/haproxy - <<< "show servers state" > /var/state/haproxy/global
```
Comment (bprashanth) [inline service-loadbalancer/haproxy_reload]: what was the problem here?

File: service-loadbalancer/template.cfg [hunk 4]
```diff
@@ -114,14 +110,14 @@ backend {{$svc.Name}}
     # http://cbonte.github.io/haproxy-dconv/configuration-1.5.html#stick-table
     stick-table type ip size 100k expire 30m
     stick on src
-    {{range $j, $ep := $svc.Ep}}server {{$ep}} {{$ep}}
+    {{range $j, $ep := $svc.Ep}}server {{$ep}} {{$ep}} check port {{$svc.BackendPort}} inter 5
```
Comment (bprashanth) [inline service-loadbalancer/template.cfg]: WDYT about doing what you once suggested (make this go straight to the liveness probe of the pod)? just a suggestion, ok to punt

**Comment by bprashanth** [comment]
Ah i think you need to update the template in unittests. While you're doing that can you please remove comments from the unittest templates, so we can modify comments in the main one without breaking tests.

File: service-loadbalancer/haproxy_reload (hunk 3 above)
Comment (aledbf) [inline service-loadbalancer/haproxy_reload]: If haproxy is reloaded too frequently the contents of that file is not correct

File: service-loadbalancer/template.cfg (hunk 4 above)
Comment (aledbf) [inline service-loadbalancer/template.cfg]: > make this go straight to the liveness probe of the pod

some apps could not define one

**Comment by aledbf** [comment]
ping @bprashanth

File: service-loadbalancer/haproxy_reload (hunk 3 above)
Comment (bprashanth) [inline service-loadbalancer/haproxy_reload]: Is that an haproxy bug? or do they document a min reloads per second? can we just limit the reloads per second to work around this?

File: service-loadbalancer/service_loadbalancer.go [hunk 5]
```diff
@@ -388,8 +391,9 @@ func (lbc *loadBalancerController) getServices() (httpSvc []service, tcpSvc []se
 				continue
 			}
 			newSvc := service{
-				Name: getServiceNameForLBRule(&s, servicePort.Port),
-				Ep:   ep,
+				Name:        getServiceNameForLBRule(&s, servicePort.Port),
+				Ep:          ep,
+				BackendPort: servicePort.TargetPort.IntVal,
```
Comment (bprashanth) [inline service-loadbalancer/service_loadbalancer.go]: Named ports are a thing, so lets just get this right here even. Please isolate the port switch case in getEndpoints to something like getTargetPort(servicePort) and call it from both places.

File: service-loadbalancer/template.cfg (hunk 4 above)
Comment (bprashanth) [inline service-loadbalancer/template.cfg]: Yeah in that case we'd default to the current 'ep:backendPort/'.

Comment (aledbf) [inline service-loadbalancer/template.cfg]: can we leave that for another PR? :)

**Comment by bprashanth** [comment]
LGTM. Would be great if you could add a test case for this under: 
https://github.com/kubernetes/kubernetes/blob/ffdfc68d11153b3ca019b84ebce8b867387bf42f/test/e2e/serviceloadbalancers.go

Sounds like all that's needed to trigger it is a failing health check on a secondary replica, so maybe just bump the replicas on the test rc and confirm all endpoints with rr algorithm.
